src/lib/error/api-error.ts

```diff
@@ -0,0 +1,270 @@
+    switch (errorKind) {
+        case 'ContentTypeError':
+            return 415;
+        case 'ValidationError':
+            return 400;
+        case 'BadDataError':
+            return 400;
+        case 'BadRequestError':
+            return 400;
+        case 'OwaspValidationError':
+            return 400;
+        case 'PasswordUndefinedError':
+            return 400;
+        case 'MinimumOneEnvironmentError':
+            return 400;
+        case 'InvalidTokenError':
+            return 401;
+        case 'NoAccessError':
+            return 403;
+        case 'UsedTokenError':
+            return 403;
+        case 'InvalidOperationError':
+            return 403;
+        case 'IncompatibleProjectError':
+            return 403;
+        case 'OperationDeniedError':
+            return 403;
+        case 'NotFoundError':
+            return 404;
+        case 'NameExistsError':
+            return 409;
+        case 'FeatureHasTagError':
+            return 409;
+        case 'RoleInUseError':
+            return 400;
+        case 'ProjectWithoutOwnerError':
+            return 409;
+        case 'UnknownError':
+            return 500;
+        case 'InternalError':
+            return 500;
+        case 'PasswordMismatch':
+        case 'PasswordMismatchError':
+            return 401;
+        case 'DisabledError':
+            return 422;
+        case 'NotImplementedError':
+            return 405;
+        case 'NoAccessError':
+            return 403;
+        case 'AuthenticationRequired':
+            return 401;
+    }
+};
+
+type ValidationErrorDescription = { description: string; path?: string };
+
+type UnleashErrorData =
+    | {
+          message: string;
+          documentationLink?: string;
+      } & (
+          | {
+                name: Exclude<
+                    UnleashApiErrorKind,
+                    | 'NoAccessError'
+                    | 'AuthenticationRequired'
+                    | 'ValidationError'
+                    | 'BadDataError'
+                    | 'BadRequestError'
+                >;
+            }
+          | {
+                name: 'NoAccessError';
+                permission: string;
+            }
+          | {
+                name: 'AuthenticationRequired';
+                path: string;
+                type: string;
+            }
+          | {
+                name:
+                    | 'ValidationError'
+                    | 'BadDataError'
+                    | 'BadRequestError'
+                    | 'MinimumOneEnvironmentError';
+                errors: [
+                    ValidationErrorDescription,
+                    ...ValidationErrorDescription[],
+                ];
+            }
+      );
+
+export class UnleashError extends Error {
+    id: string;
+
+    name: UnleashApiErrorKind;
+
+    documentationLink: string | null;
+
+    statusCode: number;
+
+    additionalParameters: object;
```

Comment:
do we have a use case for this one? I'd avoid adding speculative fields.

Reply:
We have multiple! I'd like to ultimately get rid of this myself, but we can't for now:
Some of our error messages must pass additional parameters for the Unleash app to work correctly. For instance, if the 401 you get before you log in to the UI doesn't contain `path` and `type`, you won't be redirected to login screen. Similarly, there's other error versions we have that contain extra data. 

There may well be a better way to ensure these are also passed, though, so if you have suggestions, I'm open to listening!

Reply:
Fair point